File: backend/cli_orchestrator.py
"""
CLI Orchestrator - Python implementation for managing CLI tool execution
"""

import subprocess
import asyncio
import json
import os
from typing import Dict, List, Optional, Any
from datetime import datetime
import shutil

class CLIOrchestrator:
    def __init__(self):
        self.active_processes = {}
        self.command_history = []
        self.templates = self.load_templates()
    
    def load_templates(self) -> Dict:
        """Load command templates for common workflows"""
        return {
            "claude": {
                "generateCode": {
                    "name": "Generate Code with Claude",
                    "command": "claude",
                    "args": ["code"],
                    "description": "Use Claude to generate code based on a prompt"
                },
                "reviewCode": {
                    "name": "Review Code with Claude",
                    "command": "claude",
                    "args": ["review", "{{file}}"],
                    "description": "Have Claude review a specific file"
                }
            },
            "gemini": {
                "analyze": {
                    "name": "Analyze with Gemini",
                    "command": "gemini",
                    "args": ["analyze", "{{file}}"],
                    "description": "Analyze code or requirements with Gemini"
                },
                "generateTests": {
                    "name": "Generate Tests with Gemini",
                    "command": "gemini",
                    "args": ["test", "{{file}}"],
                    "description": "Generate test cases using Gemini"
                }
            },
            "git": {
                "status": {
                    "name": "Git Status",
                    "command": "git",
                    "args": ["status"],
                    "description": "Check repository status"
                },
                "diff": {
                    "name": "Git Diff",
                    "command": "git",
                    "args": ["diff"],
                    "description": "View uncommitted changes"
                }
            }
        }
    
    def executeCommand(self, config: Dict) -> Dict:
        """Execute a CLI command synchronously"""
        command = config.get("command", "")
        args = config.get("args", [])
        stdin = config.get("stdin", "")
        cwd = config.get("cwd", os.getcwd())
        env = {**os.environ, **config.get("env", {})}
        
        command_id = f"cmd-{datetime.now().timestamp()}"
        full_command = [command] + args if args else [command]
        
        try:
            # Check if command exists
            if not shutil.which(command):
                return {
                    "commandId": command_id,
                    "command": " ".join(full_command),
                    "exitCode": 1,
                    "output": "",
                    "errorOutput": f"Command '{command}' not found",
                    "timestamp": datetime.now().isoformat()
                }
            
            # Execute command
            process = subprocess.run(
                full_command,
                input=stdin.encode() if stdin else None,
                capture_output=True,
                text=True,
                cwd=cwd,
                env=env,
                timeout=60  # 60 second timeout
            )
            
            result = {
                "commandId": command_id,
                "command": " ".join(full_command),
                "exitCode": process.returncode,
                "output": process.stdout,
                "errorOutput": process.stderr,
                "timestamp": datetime.now().isoformat()
            }
            
            self.command_history.append(result)
            return result
            
        except subprocess.TimeoutExpired:
            return {
                "commandId": command_id,
                "command": " ".join(full_command),
                "exitCode": -1,
                "output": "",
                "errorOutput": "Command timed out after 60 seconds",
                "timestamp": datetime.now().isoformat()
            }
        except Exception as e:
            return {
                "commandId": command_id,
                "command": " ".join(full_command),
                "exitCode": -1,
                "output": "",
                "errorOutput": str(e),
                "timestamp": datetime.now().isoformat()
            }
    
    async def executeCommandAsync(self, config: Dict) -> Dict:
        """Execute a CLI command asynchronously"""
        return await asyncio.to_thread(self.executeCommand, config)
    
    def executeTemplate(self, tool: str, template_name: str, variables: Dict = None) -> Dict:
        """Execute a command template with variable substitution"""
        variables = variables or {}
        template = self.templates.get(tool, {}).get(template_name)
        
        if not template:
            raise ValueError(f"Template '{template_name}' not found for tool '{tool}'")
        
        # Substitute variables
        args = []
        for arg in template.get("args", []):
            if "{{" in arg:
                for key, value in variables.items():
                    arg = arg.replace(f"{{{{{key}}}}}", str(value))
            args.append(arg)
        
        config = {
            "command": template["command"],
            "args": args,
            "stdin": variables.get("stdin", ""),
            "cwd": variables.get("projectPath", os.getcwd())
        }
        
        return self.executeCommand(config)
    
    def executeWorkflow(self, workflow_name: str, variables: Dict = None) -> List[Dict]:
        """Execute a workflow (multiple commands in sequence)"""
        # For now, return empty list - workflows would be defined separately
        return []
    
    def getHistory(self, limit: int = 50) -> List[Dict]:
        """Get command history"""
        return self.command_history[-limit:]
    
    def getTemplates(self) -> Dict:
        """Get all templates"""
        return self.templates
    
    def getAvailableTools(self) -> Dict[str, bool]:
        """Check which CLI tools are available"""
        tools = ["claude", "gemini", "git", "npm", "python", "python3", "docker"]
        availability = {}
        
        for tool in tools:
            availability[tool] = shutil.which(tool) is not None
        
        return availability

File: backend/cli/CLIOrchestrator.js
/**
 * CLI Orchestrator - Manages execution of CLI tools like Claude Code CLI, Gemini CLI, etc.
 * This replaces the complex agent orchestration with practical CLI tool integration
 */

const { spawn, exec } = require('child_process');
const path = require('path');
const fs = require('fs').promises;
const EventEmitter = require('events');

class CLIOrchestrator extends EventEmitter {
  constructor() {
    super();
    this.activeProcesses = new Map();
    this.commandHistory = [];
    this.templates = this.loadTemplates();
  }

  /**
   * Load command templates for common workflows
   */
  loadTemplates() {
    return {
      // Claude Code CLI templates
      claude: {
        generateCode: {
          name: 'Generate Code with Claude',
          command: 'claude',
          args: ['code'],
          stdin: '{{prompt}}',
          description: 'Use Claude to generate code based on a prompt'
        },
        reviewCode: {
          name: 'Review Code with Claude',
          command: 'claude',
          args: ['review', '{{file}}'],
          description: 'Have Claude review a specific file'
        },
        refactorCode: {
          name: 'Refactor with Claude',
          command: 'claude',
          args: ['refactor'],
          stdin: '{{code}}',
          description: 'Refactor code using Claude'
        }
      },
      
      // Gemini CLI templates (assuming similar interface)
      gemini: {
        analyze: {
          name: 'Analyze with Gemini',
          command: 'gemini',
          args: ['analyze', '{{file}}'],
          description: 'Analyze code or requirements with Gemini'
        },
        generateTests: {
          name: 'Generate Tests with Gemini',
          command: 'gemini',
          args: ['test', '{{file}}'],
          description: 'Generate test cases using Gemini'
        }
      },
      
      // Combined workflows
      workflows: {
        fullStackFeature: {
          name: 'Full Stack Feature Development',
          steps: [
            { tool: 'claude', template: 'generateCode', prompt: 'Create backend API: {{description}}' },
            { tool: 'claude', template: 'generateCode', prompt: 'Create frontend components: {{description}}' },
            { tool: 'gemini', template: 'generateTests' }
          ],
          description: 'Generate complete full-stack feature with tests'
        },
        codeReviewPipeline: {
          name: 'Code Review Pipeline',
          steps: [
            { tool: 'claude', template: 'reviewCode' },
            { tool: 'gemini', template: 'analyze' },
            { tool: 'claude', template: 'refactorCode', conditional: true }
          ],
          description: 'Comprehensive code review using multiple AI tools'
        }
      }
    };
  }

  /**
   * Execute a CLI command with proper handling
   */
  async executeCommand(config) {
    const { command, args = [], stdin = '', cwd = process.cwd(), env = {} } = config;
    const commandId = `cmd-${Date.now()}`;

    return new Promise((resolve, reject) => {
      try {
        // Check if command exists
        const checkCommand = process.platform === 'win32' ? 'where' : 'which';
        exec(`${checkCommand} ${command}`, (error) => {
          if (error) {
            reject(new Error(`Command '${command}' not found. Please install it first.`));
            return;
          }

          // Spawn the process
          const childProcess = spawn(command, args, {
            cwd,
            env: { ...process.env, ...env },
            shell: true
          });

          this.activeProcesses.set(commandId, childProcess);

          let output = '';
          let errorOutput = '';

          // Handle stdout
          childProcess.stdout.on('data', (data) => {
            const chunk = data.toString();
            output += chunk;
            this.emit('output', { commandId, data: chunk, stream: 'stdout' });
          });

          // Handle stderr
          childProcess.stderr.on('data', (data) => {
            const chunk = data.toString();
            errorOutput += chunk;
            this.emit('output', { commandId, data: chunk, stream: 'stderr' });
          });

          // Send stdin if provided
          if (stdin) {
            childProcess.stdin.write(stdin);
            childProcess.stdin.end();
          }

          // Handle process completion
          childProcess.on('close', (code) => {
            this.activeProcesses.delete(commandId);
            
            const result = {
              commandId,
              command: `${command} ${args.join(' ')}`,
              exitCode: code,
              output,
              errorOutput,
              timestamp: new Date().toISOString()
            };

            this.commandHistory.push(result);

            if (code === 0) {
              resolve(result);
            } else {
              reject(new Error(`Command failed with exit code ${code}: ${errorOutput}`));
            }
          });

          // Handle errors
          childProcess.on('error', (error) => {
            this.activeProcesses.delete(commandId);
            reject(error);
          });
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Execute a command template with variable substitution
   */
  async executeTemplate(tool, templateName, variables = {}) {
    const template = this.templates[tool]?.[templateName];
    if (!template) {
      throw new Error(`Template '${templateName}' not found for tool '${tool}'`);
    }

    // Substitute variables in template
    const config = {
      command: template.command,
      args: template.args?.map(arg => this.substituteVariables(arg, variables)),
      stdin: template.stdin ? this.substituteVariables(template.stdin, variables) : '',
      cwd: variables.projectPath || process.cwd()
    };

    return this.executeCommand(config);
  }

  /**
   * Execute a workflow (multiple commands in sequence)
   */
  async executeWorkflow(workflowName, variables = {}) {
    const workflow = this.templates.workflows[workflowName];
    if (!workflow) {
      throw new Error(`Workflow '${workflowName}' not found`);
    }

    const results = [];
    
    for (const step of workflow.steps) {
      try {
        // Check if step is conditional and should be skipped
        if (step.conditional && !variables[step.conditionVariable]) {
          continue;
        }

        // Prepare variables for this step
        const stepVariables = {
          ...variables,
          previousOutput: results[results.length - 1]?.output || ''
        };

        // Substitute variables in prompt if present
        if (step.prompt) {
          stepVariables.prompt = this.substituteVariables(step.prompt, stepVariables);
        }

        // Execute the step
        const result = await this.executeTemplate(step.tool, step.template, stepVariables);
        results.push(result);

        // Emit progress event
        this.emit('workflowProgress', {
          workflow: workflowName,
          step: results.length,
          totalSteps: workflow.steps.length,
          result
        });
      } catch (error) {
        this.emit('workflowError', {
          workflow: workflowName,
          step: results.length + 1,
          error: error.message
        });
        throw error;
      }
    }

    return results;
  }

  /**
   * Substitute variables in a string
   */
  substituteVariables(str, variables) {
    return str.replace(/\{\{(\w+)\}\}/g, (match, key) => {
      return variables[key] || match;
    });
  }

  /**
   * Kill a running process
   */
  killProcess(commandId) {
    const process = this.activeProcesses.get(commandId);
    if (process) {
      process.kill();
      this.activeProcesses.delete(commandId);
      return true;
    }
    return false;
  }

  /**
   * Get command history
   */
  getHistory(limit = 50) {
    return this.commandHistory.slice(-limit);
  }

  /**
   * Clear command history
   */
  clearHistory() {
    this.commandHistory = [];
  }

  /**
   * Check if a CLI tool is available
   */
  async checkToolAvailability(toolName) {
    const checkCommand = process.platform === 'win32' ? 'where' : 'which';
    
    return new Promise((resolve) => {
      exec(`${checkCommand} ${toolName}`, (error) => {
        resolve(!error);
      });
    });
  }

  /**
   * Get available tools and their status
   */
  async getAvailableTools() {
    const tools = ['claude', 'gemini', 'git', 'npm', 'python', 'docker'];
    const availability = {};

    for (const tool of tools) {
      availability[tool] = await this.checkToolAvailability(tool);
    }

    return availability;
  }

  /**
   * Save custom template
   */
  addCustomTemplate(tool, name, template) {
    if (!this.templates[tool]) {
      this.templates[tool] = {};
    }
    this.templates[tool][name] = template;
  }

  /**
   * Get all templates
   */
  getTemplates() {
    return this.templates;
  }
}

module.exports = CLIOrchestrator;